### ISSUE TITLE
Build and install Coq librairies in -vos mode or in only-parsing mode.

### ISSUE THREAD

**Issue by Zimmi48** [coq/coq#13648]
[Recently](https://github.com/coq/coq/wiki/Coq-Call-2020-11-25), we have discussed having a more lightweight CI run on PRs by default (with the full CI still being required before merging a PR). While the easiest way to make a lighter CI is of course to reduce the number of jobs it contains, we can also consider doing some more lightweight static analyses on external projects that would inform us of part of the breakage while being faster to run.

I would thus be interested in:

- An only-parsing compilation and installation mode: all files would be parsed and only commands related to this phase (notations, imports) would be evaluated. This mode would be useful to discover notation issues without having to fully build the tested projects. It would also be useful to evaluate if we can remove some particular syntactic form of a command or a tactic (or if we can remove entirely a given command or tactic).

- A `-vos` compilation and installation mode. I thought it would be easy to do since the `-vos` infrastructure already exists. But it turns out that you cannot install a library that you have compiled in `-vos` mode because the `.vo` and `.glob` files are not there. It also turns out that switching tested projects to `-vos` mode is very dependent on their build infrastructure. It makes me want to come back to @gares's proposal to have `-vos` also generate `.vo` files instead of `.vos` files (and same thing for the only-parsing mode above). The only difference would be that the `.vo` files would be "incomplete".

**Comment by ejgallego** [comment]
That's an interesting idea, however I'm not sure the current infrastructure will make doing this easy or worthwhile as you point out.

You can already use Coq in only-parsing mode with `sercomp`, but be aware that this doesn't do exactly what you want due to the way the STM / require do work. `-vos` and incomplete stuff won't work likely work too well for CI due to the way deps are tracked.

The document / project manager I've been working on does allow both setups and indeed allows for persistance, however a lot of work is needed so this is not realistic. In fact it doesn't really distinguish anymore between parsing/vos/vo , but works in an incremental way.

IMHO more important for CI is to be incremental, which could be done relatively easy using Dune cache + dune for some projects, only few bits are missing there.

**Comment by Zimmi48** [comment]
> IMHO more important for CI is to be incremental, which could be done relatively easy using Dune cache + dune for some projects, only few bits are missing there.

I agree with this but won't the whole Coq Dune cache be invalidated if the OCaml code of Coq changes? In this case, this will not improve things a lot for most PRs (even if it's still worthwhile to do).

**Comment by gares** [comment]
> I would thus be interested in:
> 
>     * An only-parsing compilation and installation mode: all files would be parsed and only commands related to this phase (notations, imports) would be evaluated. This mode would be useful to discover notation issues without having to fully build the tested projects. It would also be useful to evaluate if we can remove some particular syntactic form of a command or a tactic (or if we can remove entirely a given command or tactic).

This is really hard, I'd love it to be possible, but the parsing and execution phase are still entangled.
Eg `Import` changes the parser but the only way we have to see that is by executing if fully.

>     * A `-vos` compilation and installation mode. I thought it would be easy to do since the `-vos` infrastructure already exists. But it turns out that you cannot install a library that you have compiled in `-vos` mode because the `.vo` and `.glob` files are not there. It also turns out that switching tested projects to `-vos` mode is very dependent on their build infrastructure. It makes me want to come back to @gares's proposal to have `-vos` also generate `.vo` files instead of `.vos` files (and same thing for the only-parsing mode above). The only difference would be that the `.vo` files would be "incomplete".

I had no time to play with the idea, but I still believe  my proposal to use dune profiles to feed coqc options is doable IMO.
The v -> vio/vos part has the same rules of the v -> vo one (if we stop naming the output file differently).


Said that, by lack of systematic `Proof using` annotations and the entangling of parsing with execution, I'm not so sure this "check" can be much faster (I mean, if it is not 10x faster it may not be worth it).

**Comment by Zimmi48** [comment]
>  Eg `Import` changes the parser but the only way we have to see that is by executing if fully.

Yes, I know it does, but I was thinking we could limit the import operation to the parsing rules if we limit what we store in the `vo`.

> Said that, by lack of systematic `Proof using` annotations and the entangling of parsing with execution, I'm not so sure this "check" can be much faster (I mean, if it is not 10x faster it may not be worth it).

Sure, but it would be easy enough to add those.

**Comment by ejgallego** [comment]
> I agree with this but won't the whole Coq Dune cache be invalidated if the OCaml code of Coq changes? In this case, this will not improve things a lot for most PRs (even if it's still worthwhile to do).

Will help a lot in some cases, little on others indeed.

**Comment by ejgallego** [comment]
We can easily try and see what the speeups are for parsing only using `sercomp`, but the problem is that it doesn't do a lot if you don't generate at least the `.vos` / `.vo` right?

**Comment by gares** [comment]
> Yes, I know it does, but I was thinking we could limit the import operation to the parsing rules if we limit what we store in the `vo`.

I thought a bit about that with Maxime, it's doable but not easy. For example `Module Import M := F(X)` loads the parsing rules of the functor F applied to X. One would need to (also) clearly separate functor application into two steps (the parsing effects and the rest). And in order to do anything with `F(X)` you need to at least interpret the nametab part of `Import` (otherwise you don't know what `F` is). I think it is still an objective to seek, but it requires many cleanups.
